crates/client/settlement/src/sync_state.rs

```diff
@@ -0,0 +1,207 @@
+        Ok((starknet_block, substrate_block_hash))
+    }
+
+    fn verify_starknet_state(substrate_client: &SC, state: &StarknetState) -> anyhow::Result<()> {
+        let level: u64 = state.block_number.try_into()?;
+        let (block, _) = Self::get_starknet_block(substrate_client, level)?;
+
+        // Verify that current onchain state is consistent with corresponding Madara block
+        if state.state_root != block.header().global_state_root {
+            return Err(UnrecoverableError::StateRootMismatch(level).into());
+        }
+
+        Ok(())
+    }
+
+    async fn get_starknet_state<SP>(settlement_provider: &SP) -> anyhow::Result<StarknetState>
+    where
+        SP: ?Sized + SettlementProvider,
+    {
+        let last_settled_state = match settlement_provider.get_state().await? {
+            Some(state) => state,
+            None => {
+                log::warn!("[settlement] State is not initialized");
+                // TODO: we skip this check for now, to simplify the setup (allow non-initialized state)
+                // We use zero state as a starting point (note that genesis state root is not zero actually)
+                return Ok(StarknetState::default());
+            }
+        };
+
+        Ok(last_settled_state)
+    }
+
+    async fn update_starknet_state<SP>(
+        substrate_client: &SC,
+        settlement_provider: &SP,
+        prev_state: &StarknetState,
+        next_block: &StarknetBlock,
+        substrate_block_hash: B::Hash,
+    ) -> anyhow::Result<StarknetState>
+    where
+        SP: ?Sized + SettlementProvider,
+    {
+        let next_state = StarknetState {
+            block_number: next_block.header().block_number.into(),
+            state_root: next_block.header().global_state_root,
+            config_hash: prev_state.config_hash,
+            chain_id: prev_state.chain_id,
+        };
+
+        let mut messages_to_l1: Vec<MessageL2ToL1> = Vec::new();
+        let mut messages_to_l2: Vec<MessageL1ToL2> = Vec::new();
+
+        for tx in next_block.transactions() {
+            if let Transaction::L1Handler(l1_handler) = tx {
+                messages_to_l2.push(l1_handler.clone().into());
+            }
+            let tx_hash = TransactionHash(tx.compute_hash::<H>(next_state.chain_id.into(), false).into());
+            substrate_client
+                .runtime_api()
+                .get_tx_messages_to_l1(substrate_block_hash, tx_hash)?
+                .into_iter()
+                .for_each(|msg| messages_to_l1.push(msg.into()));
+        }
+
+        let program_output = StarknetOsOutput {
+            prev_state_root: prev_state.state_root,
+            new_state_root: next_state.state_root,
+            block_number: next_state.block_number,
+            block_hash: next_block.header().hash::<H>().into(),
+            config_hash: next_state.config_hash,
+            messages_to_l1,
+            messages_to_l2,
+        };
```

Comment:
We are short-circuiting the SNOS run here, isn't it?

Can you add a comment about it please somewhere